include/tvm/relax/dataflow_analysis.h

```diff
@@ -0,0 +1,198 @@
+   * (i.e., the function arguments). */
+  Array<Var> args;
+
+  /*! \brief Index of the binding block in the SeqExpr where the binding is found.
+   *  Convention: We put the SeqExpr body at one block past the final block. */
+  size_t block_idx;
+
+  /*! \brief Index of the binding within the binding block corresponding to this binding.
+   *  Convention: Both the If condition and merge are mapped to the same index.
+   *  We use the kind to distinguish. */
+  size_t binding_idx;
+
+  /*! \brief The kind of binding this is. */
+  BindingNodeKind kind;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("seq", &seq);
+    v->Visit("args", &args);
+    v->Visit("block_idx", &block_idx);
+    v->Visit("binding_idx", &binding_idx);
+    v->Visit("kind", &kind);
+  }
+
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const char* _type_key = "relax.analysis.GraphBinding";
+  TVM_DECLARE_BASE_OBJECT_INFO(GraphBindingNode, Object);
+};
+
+/*! \brief Representation of a binding in the control flow graph */
+class GraphBinding : public ObjectRef {
+ public:
+  /*!
+   * \brief Create a GraphBinding. See the docs on GraphBindingNode for further details.
+   *
+   * \param seq: The SeqExpr in which the binding resides.
+   * \param args: The arguments to the binding (only nonempty for the first binding:
+   *   these will be the function arguments)
+   * \param block_idx: The index of the BindingBlock in the SeqExpr
+   *   where the binding resides (for the return expression, use one past the final block).
+   * \param binding_idx: The index of the binding in the BindingBlock corresponding to the binding.
+   * \param kind: The kind of binding this is. (Used especially to distinguish If node conditions
+   * from the merge after the If)
+   */
+  TVM_DLL static GraphBinding Create(const SeqExpr& seq, const Array<Var>& args, size_t block_idx,
+                                     size_t binding_idx, BindingNodeKind kind);
+
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(GraphBinding, ObjectRef, GraphBindingNode);
+};
+
+/* A control flow graph corresponding to a function.
+ */
+class ControlFlowGraphNode : public Object {
+ public:
+  /*! \brief The bindings in the graph. 0 is the entry point. */
+  Array<GraphBinding> bindings;
+  /*! \brief The ith member is the list of predecessors (indices) to binding i in bindings. */
+  Array<Array<Integer>> preds;
+  /*! \brief The ith member is the list of successors (indices) to binding i in bindings. */
+  Array<Array<Integer>> succs;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("bindings", &bindings);
+    v->Visit("preds", &preds);
+    v->Visit("succs", &succs);
+  }
+
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const char* _type_key = "relax.analysis.ControlFlowGraph";
+  TVM_DECLARE_BASE_OBJECT_INFO(ControlFlowGraphNode, Object);
+};
+
+class ControlFlowGraph : public ObjectRef {
```

Comment:
Looking at this structure, I think it could be generalized to also represent a data-dependency graph, and most of the functionality would also carry over.

* Both the predecessors in a control-flow graph are analogous to the inputs for a data-dependency graph, and both could be represented by `Array<Var>`.
* Both the successors in a control-flow graph are analogous to the outputs of a data-dependency graph, and both could be represented by `Array<Var>`.
* The `DataflowAnalysis` function would operate identically in both cases, either flowing things that are known at a specific time for a control-flow graph, or flowing things that are known about a specific value for a data-dependency graph.

What are your thoughts on generalizing the utility?  I think the main drawback would be if there's a fundamental assumption made about the graph structure that only holds for one of the two, but they look like they might be similar enough to have lots of overlap.

Reply:
I think a dependency graph is more specific than a CFG, so I would have to think about whether the same analyses would work on one. It's worth further thought.